tests/unit/states/test_file.py

```diff
@@ -2029,3 +2033,67 @@ def test__bad_input(self):
             'comment': '/bad-directory-name/ does not exist or is not a directory.',
         }
         assert filestate.tidied(name='/bad-directory-name/') == exp
+
+
+class TestFilePrivateFunctions(TestCase, LoaderModuleMockMixin):
+    def setup_loader_modules(self):
+        return {
+            filestate: {
+                '__salt__': {'file.stats': filemod.stats},
+            }
+        }
+
+    @destructiveTest
+    @skipIf(salt.utils.platform.is_windows(), 'File modes do not exist on windows')
+    def test__check_directory(self):
+        '''
+        Test the _check_directory function
+        Make sure that recursive file permission checks return correctly
+        '''
+        # set file permissions
+        # Run _check_directory function
+        # Verify that it returns correctly
+        # Delete tmp directory structure
+        root_tmp_dir = os.path.join(TMP, 'test__check_dir')
+        depth = 3
+        try:
+            def create_files(tmp_dir):
+                for f in range(depth):
+                    path = os.path.join(tmp_dir, 'file_{:03}.txt'.format(f))
+                    with salt.utils.files.fopen(path, 'w+'):
+                        os.chmod(path, 0o777)
+
+            # Create tmp directory structure
+            os.mkdir(root_tmp_dir)
+            os.chmod(root_tmp_dir, 0o777)
+            create_files(root_tmp_dir)
+
+            for d in range(depth):
+                dir_name = os.path.join(root_tmp_dir, 'dir{:03}'.format(d))
+                os.mkdir(dir_name)
+                os.chmod(dir_name, 0o777)
+                create_files(dir_name)
+                for s in range(depth):
+                    sub_dir_name = os.path.join(dir_name, 'dir{:03}'.format(s))
+                    os.mkdir(sub_dir_name)
+                    os.chmod(sub_dir_name, 0o777)
+                    create_files(sub_dir_name)
+
+            # Set some bad permissions
+            changed_files = {
+                os.path.join(root_tmp_dir, 'file_000.txt'),
+                os.path.join(root_tmp_dir, 'dir002', 'file_000.txt'),
+                os.path.join(root_tmp_dir, 'dir000', 'dir001', 'file_002.txt'),
+                os.path.join(root_tmp_dir, 'dir001', 'dir002'),
+                os.path.join(root_tmp_dir, 'dir002', 'dir000'),
+                os.path.join(root_tmp_dir, 'dir001'),
+            }
+            for c in changed_files:
+                os.chmod(c, 0o770)
+
+            ret = filestate._check_directory(root_tmp_dir, dir_mode=0o777, file_mode=0o777, recurse=['mode'])
```

Comment:
I'm sure the test is failing because the options here are incorrect, what should we be putting in for options?

Reply:
@amendlik Can you verify the options we're passing here? It's currently returning that every file is going to be changed. I would expect that it's only supposed to return the files that are actually having their mode changed. Is that not the case?

Reply:
I see 2 problems here:
1. You need to pass the `mode` arguments as strings, as they would be coming from a YAML file. So the call to `_check_directory` should look like this:
```
ret = filestate._check_directory(root_tmp_dir, dir_mode='0777', file_mode='0777', recurse=['mode'])
```
2. This will still fail under Python 3 unless you have the changes from #52211. I'm not sure why those are not yet on this branch, but you'll have to merge those in before this test will pass.

Reply:
@amendlik I changed it to strings and now the test is passing :+1:  I will also look at cherry-picking  or merging #52211 into 2019.2.1.  @twangboy, should that be a new PR?  Add the merge label, or could it be grouped with this one?